[ContactUs] 4, 5, 6, 7 e 8

bruunomooura opened the pull request:
Peço desculpas pelo tamanho deste PR.
Muitos ajustes já foram feitos anteriormente nas três primeiras tarefas, então decidi agrupar as alterações restantes, achando que ficariam menores. No entanto, acabei fazendo algumas modificações na organização do projeto, o que resultou na inclusão de mais arquivos do que o esperado.
O último commit é maior, pois envolve a implementação da ViewModel.

bruunomooura committed:
Completed removal of object orientation in the MelContactUs flow.

bruunomooura committed:
[ContactUs] 5 - Utilizar protocolos e DIP - Completed
[ContactUs] 6 - Remover Singletons - Completed

bruunomooura committed:
[ContactUs] 7 - Tratar retain cycles - Completed

bruunomooura committed:
[ContactUs] 8 - Criar ViewModel - Completed

ppedroam commented on CleanCodeApp/Modules/Features/Mel/ContactsUs/MelContactUsService.swift:
```diff
@@ -7,10 +7,22 @@
 
-final class MelContactUsService {
+protocol MelContactUsServiceProtocol: AnyObject {
+    func fetchContactData(completion: @escaping (Result<ContactUsModel, Error>) -> Void)
```
usar async await

ppedroam commented on CleanCodeApp/Modules/Features/Mel/ContactsUs/View/MelContactUsViewController.swift:
```diff
@@ -0,0 +1,107 @@
+    private let contactUsService: MelContactUsServiceProtocol
+    private let melLoadingView: MelLoadingViewProtocol
+    
+    init(viewModel: MelContactUsViewModelProtocol = MelContactUsViewModel(appOpener: ExternalAppOpener(application: UIApplication.shared)),
```
nao fica muito limpo o init quando vc passa um valor default enquanto injeta um objeto nesse valor default
sugiro nao colocar valor default no caso
pq caso aumente o numero de dependencias no valor default, ele só vai crescer
ai passa tudo isso para a factory desse viewcontroller

ppedroam commented on CleanCodeApp/Modules/Features/Mel/ContactsUs/View/MelContactUsViewController.swift:
```diff
@@ -0,0 +1,107 @@
+    }
+    
+    private func displaySuccessAlert() {
+        Globals.alertMessage(title: MelContactUsStrings.successAlertTitle.rawValue,
```
usar composição para mostrar alerta
nao usar metodo global estatica
criar protocolo de alerta para sua viewController e colocar implementação padrão desse protocolo

ppedroam commented on CleanCodeApp/Modules/Features/Mel/ContactsUs/ViewModel/MelContactUsViewModel.swift:
```diff
@@ -0,0 +1,137 @@
+    func sendMessageToSupport(message: String) {
+        guard !message.isEmpty, let email = contactModel?.mail else { return }
+        delegate?.presentLoading()
+        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { [weak self] in
```
remover essa logica do loading
ja que ele vai ser escondido na resposta da chamada da service

ppedroam commented on CleanCodeApp/Modules/Features/Mel/MelContactUrlFactory.swift:
```diff
@@ -0,0 +1,25 @@
+    static func makeEmailUrl(for email: String) -> ExternalUrlCreator {
+        return EmailUrlCreator(email: email)
+    }
+}
```
muito bom essa organização

ppedroam reviewed (changes requested):
code review 1

bruunomooura committed:
Adjustments made in code review completed.

bruunomooura commented:
@ppedroam ajustes realizados.

bruunomooura committed:
Small adjustments.

ppedroam commented on CleanCodeApp/Modules/Features/Mel/ContactsUs/MelContactUsService.swift:
```diff
@@ -8,40 +8,32 @@
-                completion(.failure(error))
-            }
-        }
+        let data = try await networking.request(url, method: .get, parameters: nil, headers: nil)
```
ola o tanto que o codigo fica mais limpo com async await

ppedroam commented on CleanCodeApp/Modules/Features/Mel/MelAlertDisplay.swift:
```diff
@@ -0,0 +1,47 @@
+
+protocol MelAlertDisplayProtocol: AnyObject {
+    func displaySuccessAlert(on viewController: UIViewController, completion: (() -> Void)?)
+    func displayErrorAlert(on viewController: UIViewController, mustDismiss: Bool)
```
nao é legal performar acoes na viewController fora dela
isso quebra a coesão do objeto já que ele vai ser responsavel pelos próprios comportamentos mas tambem outros objetos vão ter comportamentos sobre essa controller
logo nao passa a viewController de parametro

te recomendo usar um Where Self: UIViewController no protocolo
e criar uma implementacao padrão para esse protocolo
aí vc n precisaria injetar esse objeto na controller, mas sim fazer sua controller extender esse protocolo, que é a ideia de composição

ppedroam commented on CleanCodeApp/AppDelegate/SceneDelegate.swift:
```diff
@@ -17,7 +17,7 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         
         self.window = UIWindow(frame: UIScreen.main.bounds)
-        let rootViewController = getRootViewController(forUser: .brunoMoura)
+        let rootViewController = HomeFactory3.make()//getRootViewController(forUser: .jorgeRoberto)
```
pode deixar a sua linha mesmo

ppedroam reviewed (changes requested):
code review 2

bruunomooura committed:
Merge branch 'ppedroam:main' into main

bruunomooura committed:
- Added default implementation to the `MelAlertDisplayProtocol` protocol with `where Self: UIViewController` restriction;
- Removed passing `UIViewController` as a parameter in alert display methods;
- Conformed `MelContactUsViewController` to the `MelAlertDisplayProtocol` protocol to use the default implementation;
- Removed the injection of `MelAlertDisplay` in the `MelContactUsViewController` initializer.

bruunomooura commented:
@ppedroam ajustes concluídos.
